### ISSUE TITLE
.concat returning invalid length when concating several empty arrays

### ISSUE THREAD

**Issue by ScottWeinstein** [fholm/IronJS#39]
The fix lets the code run w/o errors, that said, the results aren't correct yet.

Compiling this CoffeeScript fragment `foo -> 1` with IronJS gives:

``` javascript
var ;
 foo(function() {
var ;
return 1;
});
```

but the correct result should look like

``` javascript
foo(function() {
  return 1;
});
```

I used the following code to test

```

#light
#r @"C:\Users\Scott\Documents\d\GitHub\IronJS\Src\IronJS\bin\Release\IronJS.dll"

open System.Net
open IronJS
open IronJS.Hosting.FSharp

let cssrcUri = System.Uri "http://jashkenas.github.com/coffee-script/extras/coffee-script.js"
let cssrcPath = "C:\Users\Scott\Documents\d\GitHub\CoffeeScriptIronJS\FSCoffeeScriptIronJS\coffee-script.js"
let coffeescriptCompilerSrc =
    if System.IO.File.Exists(cssrcPath) then
     System.IO.File.ReadAllText(cssrcPath)
    else
        let wc = new WebClient()
        wc.DownloadString cssrcUri


let ctx = IronJS.Hosting.FSharp.createContext()
let ret1 = ctx |> execute(coffeescriptCompilerSrc)
let ret2 = ctx |> execute("CoffeeScript.compile('foo -> 1', {bare: true});")
printf "%s", ret2
```

**Comment by fholm** [comment]
I've finally gotten through my two crunch-weeks at work and got time to get back working on IronJS, I'll look into this ASAP and make sure we fix it within the next few days.

**Comment by otac0n** [comment]
So, the source of coffeescript says this:

```
if scope.expressions is this
      if o.scope.hasDeclarations()
        code += "#{@tab}var #{ scope.declaredVariables().join(', ') };\n"
```

The issue is certainly that `o.scope.hasDeclarations()` is returning a false positive.  (not that there is a bug in CoffeeScript...)

The definitions of these in the coffeestript source is something like:

```
a.prototype.hasDeclarations = function () {
    return !!this.declaredVariables().length
}, a.prototype.declaredVariables = function () {
    var a, b, c, d, e, f;
    a = [], b = [], f = this.variables;
    for (d = 0, e = f.length; d < e; d++) c = f[d], c.type === "var" && (c.name.charAt(0) === "_" ? b : a).push(c.name);
        return a.sort().concat(b.sort())
}
```

My bet is on `sort` or `concat` causing issues.  Not sure yet.

**Comment by fholm** [comment]
Thanks for tracking this down @otac0n, i'm suspecting that there's something going on in that for-loop that isn't kosher with the IronJS parser (just saying, the bug is obviously in IronJS no doubt). 

I'm betting either on `sort` and `concat` or some of the logic inside the loop.

**Comment by otac0n** [comment]
```
var a = [];
var b = [];
a.concat(b).length;
```

returns 1.

At this point, I'm almost certain that it is the `concat` function.

**Comment by fholm** [comment]
This problem is fixed in 34584aca8643ba250bf79588708de50f143b16ca.
@otac0n: Thanks for locating the source function, finding the exact bug was easy once I knew where to look.

**Comment by ScottWeinstein** [comment]
awesome, thanks!
